File: Common/MultiTextWriter.vb
Public Class MultiTextWriter
    Inherits IO.TextWriter

    Public Sub New(ByVal TextWriters As IEnumerable(Of IO.TextWriter))
        Me.Bases = TextWriters.ToArray()
    End Sub

    Public Sub New(ParamArray ByVal TextWriters As IO.TextWriter())
        Me.New(DirectCast(TextWriters, IEnumerable(Of IO.TextWriter)))
    End Sub

#Region "Writes Group"
    Public Overrides Sub Write(value As String)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(buffer() As Char, index As Integer, count As Integer)
        For Each B In Me.Bases
            B.Write(buffer, index, count)
        Next
    End Sub

    Public Overrides Sub Write(value As Char)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(buffer() As Char)
        For Each B In Me.Bases
            B.Write(buffer)
        Next
    End Sub

    Public Overrides Sub Write(format As String, arg0 As Object)
        For Each B In Me.Bases
            B.Write(format, arg0)
        Next
    End Sub

    Public Overrides Sub Write(format As String, arg0 As Object, arg1 As Object)
        For Each B In Me.Bases
            B.Write(format, arg0, arg1)
        Next
    End Sub

    Public Overrides Sub Write(format As String, arg0 As Object, arg1 As Object, arg2 As Object)
        For Each B In Me.Bases
            B.Write(format, arg0, arg1, arg2)
        Next
    End Sub

    Public Overrides Sub Write(format As String, ParamArray arg() As Object)
        For Each B In Me.Bases
            B.Write(format, arg)
        Next
    End Sub

    Public Overrides Sub Write(value As Boolean)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As Decimal)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As Double)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As Integer)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As Long)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As Object)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As Single)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As UInteger)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub

    Public Overrides Sub Write(value As ULong)
        For Each B In Me.Bases
            B.Write(value)
        Next
    End Sub
#End Region

#Region "WriteLines Group"
    Public Overrides Sub WriteLine()
        For Each B In Me.Bases
            B.WriteLine()
        Next
    End Sub

    Public Overrides Sub WriteLine(buffer() As Char)
        For Each B In Me.Bases
            B.WriteLine(buffer)
        Next
    End Sub

    Public Overrides Sub WriteLine(buffer() As Char, index As Integer, count As Integer)
        For Each B In Me.Bases
            B.WriteLine(buffer, index, count)
        Next
    End Sub

    Public Overrides Sub WriteLine(format As String, arg0 As Object)
        For Each B In Me.Bases
            B.WriteLine(format, arg0)
        Next
    End Sub

    Public Overrides Sub WriteLine(format As String, arg0 As Object, arg1 As Object)
        For Each B In Me.Bases
            B.WriteLine(format, arg0, arg1)
        Next
    End Sub

    Public Overrides Sub WriteLine(format As String, arg0 As Object, arg1 As Object, arg2 As Object)
        For Each B In Me.Bases
            B.WriteLine(format, arg0, arg1, arg2)
        Next
    End Sub

    Public Overrides Sub WriteLine(format As String, ParamArray arg() As Object)
        For Each B In Me.Bases
            B.WriteLine(format, arg)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Boolean)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Char)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Decimal)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Double)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Integer)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Long)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Object)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As Single)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As String)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As UInteger)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub

    Public Overrides Sub WriteLine(value As ULong)
        For Each B In Me.Bases
            B.WriteLine(value)
        Next
    End Sub
#End Region

#Region "Asyncs Group"
    Public Overrides Function WriteAsync(buffer() As Char, index As Integer, count As Integer) As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteAsync(buffer, index, count)))
    End Function

    Public Overrides Function WriteAsync(value As Char) As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteAsync(value)))
    End Function

    Public Overrides Function WriteAsync(value As String) As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteAsync(value)))
    End Function

    Public Overrides Function WriteLineAsync() As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteLineAsync()))
    End Function

    Public Overrides Function WriteLineAsync(buffer() As Char, index As Integer, count As Integer) As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteLineAsync(buffer, index, count)))
    End Function

    Public Overrides Function WriteLineAsync(value As Char) As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteLineAsync(value)))
    End Function

    Public Overrides Function WriteLineAsync(value As String) As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.WriteLineAsync(value)))
    End Function
#End Region

    Public Overrides Sub Close()
        For Each B In Me.Bases
            B.Close()
        Next
    End Sub

    Public Overrides Function CreateObjRef(requestedType As Type) As Runtime.Remoting.ObjRef
        Throw New NotSupportedException()
    End Function

    Protected Overrides Sub Dispose(disposing As Boolean)
        For Each B In Me.Bases
            B.Dispose()
        Next
    End Sub

    Public Overrides Sub Flush()
        For Each B In Me.Bases
            B.Flush()
        Next
    End Sub

    Public Overrides Function FlushAsync() As Task
        Return Task.WhenAll(Me.Bases.Select(Function(B) B.FlushAsync()))
    End Function

    Public Overrides Function InitializeLifetimeService() As Object
        Throw New NotSupportedException()
    End Function

    Public Overrides Function ToString() As String
        Dim Res = New Text.StringBuilder()
        Res.Append(NameOf(MultiTextWriter)).Append("{"c)

        Dim Bl = False
        For Each B In Me.Bases
            If Bl Then
                Res.Append(","c)
            End If
            Bl = True
            Res.Append(B.ToString())
        Next

        Res.Append("}"c)
        Return Res.ToString()
    End Function

    Public Overrides ReadOnly Property Encoding As Text.Encoding
        Get
            Return Me.Bases.FirstOrDefault()?.Encoding
        End Get
    End Property

    Public Overrides Property NewLine As String
        Get
            Return Me.Bases.FirstOrDefault()?.NewLine
        End Get
        Set(value As String)
            For Each B In Me.Bases
                B.NewLine = value
            Next
        End Set
    End Property

    Public Overrides ReadOnly Property FormatProvider As IFormatProvider
        Get
            Return Me.Bases.FirstOrDefault()?.FormatProvider
        End Get
    End Property

    Private ReadOnly Bases As IO.TextWriter()

End Class

File: Common/MultiTextWriter.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;

namespace Ks.Common
{
    public class MultiTextWriter : System.IO.TextWriter
    {
        public MultiTextWriter(IEnumerable<System.IO.TextWriter> TextWriters)
        {
            this.Bases = TextWriters.ToArray();
        }

        public MultiTextWriter(params System.IO.TextWriter[] TextWriters) : this((IEnumerable<System.IO.TextWriter>) TextWriters)
        {
        }

        public override void Write(string value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(char[] buffer, int index, int count)
        {
            foreach (var B in this.Bases)
            {
                B.Write(buffer, index, count);
            }
        }

        public override void Write(char value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(char[] buffer)
        {
            foreach (var B in this.Bases)
            {
                B.Write(buffer);
            }
        }

        public override void Write(string format, object arg0)
        {
            foreach (var B in this.Bases)
            {
                B.Write(format, arg0);
            }
        }

        public override void Write(string format, object arg0, object arg1)
        {
            foreach (var B in this.Bases)
            {
                B.Write(format, arg0, arg1);
            }
        }

        public override void Write(string format, object arg0, object arg1, object arg2)
        {
            foreach (var B in this.Bases)
            {
                B.Write(format, arg0, arg1, arg2);
            }
        }

        public override void Write(string format, params object[] arg)
        {
            foreach (var B in this.Bases)
            {
                B.Write(format, arg);
            }
        }

        public override void Write(bool value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(decimal value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(double value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(int value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(long value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(object value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(float value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(uint value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void Write(ulong value)
        {
            foreach (var B in this.Bases)
            {
                B.Write(value);
            }
        }

        public override void WriteLine()
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine();
            }
        }

        public override void WriteLine(char[] buffer)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(buffer);
            }
        }

        public override void WriteLine(char[] buffer, int index, int count)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(buffer, index, count);
            }
        }

        public override void WriteLine(string format, object arg0)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(format, arg0);
            }
        }

        public override void WriteLine(string format, object arg0, object arg1)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(format, arg0, arg1);
            }
        }

        public override void WriteLine(string format, object arg0, object arg1, object arg2)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(format, arg0, arg1, arg2);
            }
        }

        public override void WriteLine(string format, params object[] arg)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(format, arg);
            }
        }

        public override void WriteLine(bool value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(char value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(decimal value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(double value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(int value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(long value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(object value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(float value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(string value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(uint value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override void WriteLine(ulong value)
        {
            foreach (var B in this.Bases)
            {
                B.WriteLine(value);
            }
        }

        public override Task WriteAsync(char[] buffer, int index, int count)
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteAsync(buffer, index, count)));
        }

        public override Task WriteAsync(char value)
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteAsync(value)));
        }

        public override Task WriteAsync(string value)
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteAsync(value)));
        }

        public override Task WriteLineAsync()
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteLineAsync()));
        }

        public override Task WriteLineAsync(char[] buffer, int index, int count)
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteLineAsync(buffer, index, count)));
        }

        public override Task WriteLineAsync(char value)
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteLineAsync(value)));
        }

        public override Task WriteLineAsync(string value)
        {
            return Task.WhenAll(this.Bases.Select(B => B.WriteLineAsync(value)));
        }

        public override void Close()
        {
            foreach (var B in this.Bases)
            {
                B.Close();
            }
        }

        public override System.Runtime.Remoting.ObjRef CreateObjRef(Type requestedType)
        {
            throw new NotSupportedException();
        }

        protected override void Dispose(bool disposing)
        {
            foreach (var B in this.Bases)
            {
                B.Dispose();
            }
        }

        public override void Flush()
        {
            foreach (var B in this.Bases)
            {
                B.Flush();
            }
        }

        public override Task FlushAsync()
        {
            return Task.WhenAll(this.Bases.Select(B => B.FlushAsync()));
        }

        public override object InitializeLifetimeService()
        {
            throw new NotSupportedException();
        }

        public override string ToString()
        {
            var Res = new System.Text.StringBuilder();
            Res.Append(nameof(MultiTextWriter)).Append('{');

            var Bl = false;
            foreach (var B in this.Bases)
            {
                if (Bl)
                {
                    Res.Append(',');
                }

                Bl = true;
                Res.Append(B.ToString());
            }

            Res.Append('}');
            return Res.ToString();
        }

        public override System.Text.Encoding Encoding => this.Bases.FirstOrDefault()?.Encoding;

        public override string NewLine
        {
            get => this.Bases.FirstOrDefault()?.NewLine;
            set
            {
                foreach (var B in this.Bases)
                {
                    B.NewLine = value;
                }
            }
        }

        public override IFormatProvider FormatProvider => this.Bases.FirstOrDefault()?.FormatProvider;

        private readonly System.IO.TextWriter[] Bases;
    }
}
